[snap] Add 'snap/snapcraft.yaml'

bboymimi opened the pull request:
A snap is a bundle of your app and its dependencies that works without
modification across Linux. Snaps are discoverable and installable from
the Snap store.

Signed-off-by: Gavin Guo <gavin.guo@canonical.com>

masatake commented:
Thank you, but I don't understand what is this. 
Could you write how to utilize the yaml file and benefits to snap/README.rst or somewhere?

bboymimi commented:
Hi Masatake,

The snap/ folder is only for the snap related files. I've updated the commit message and make the snap explanation more clearly. Could you help to review the commit and see if there are any problems? Thank you!

masatake commented:
Thank you for updating.
I have a question about snap. Is it possible to distribute man pages?
I introduced some incompatiblities to u-ctags. I have written them in our man pages.
I would like users to know the incompatibilities.

bboymimi commented:
Hi Masatake,

https://forum.snapcraft.io/t/support-for-man-pages/2299
According to the above discussion, the 'man' feature still is not available. We can add the man page once it's supported.

masatake commented:
I found one critical bad thing.
`universal-ctags.ctags`, this is really bad name. Universal-ctags considers the extension `.ctags` is for defining a command line based parser. (I called it optlib in our manuals.)
You will see man .ctags file in https://github.com/universal-ctags/ctags/tree/master/optlib

Is there anyway not to use `universal-ctags.ctags`?

masatake commented:
... Though I merged, I would like to revert this.
You can distribute the snapfile at another repository like https://github.com/universal-ctags/ctags-docker instead. How do you think being a member of universal-ctags organization to maintain the snapfile?

I'm very sorry to be late to propose this. I'm being busy so I could not take much time for this project now.

masatake commented:
Your excellent commit log can be convert to README.md of the new repository.

bboymimi commented:
Hi Masatake,

It's ok, we can separate the snap package version to another repo.
Please let me know how to be a member of the ctags organization. And 
how to create a separate repo under your main repo 'universal-ctags.'

Besides, we need to settle down the name of the executable file. How 
about to change the name to 'universal-ctags'? In my perspective, if
we set up the name to 'ctags' only, it will be a conflict between
different implementations of ctags, like 'exuberant ctags' if the user
installs multiple 'ctags' implementation at the same time.

As described in my commit message, we can then set up the alias when
there are multiple 'ctags' implementations exist in the system by: 

$ sudo snap alias universal-ctags ctags
Added:
  - universal-ctags as ctags
  
$ ls /snap/bin/ctags -al
lrwxrwxrwx 1 root root 15 Dec 11 02:50 /snap/bin/ctags -> universal-ctags

I also updated the branch for the naming rule. Please have a look and
let me know if there is anything need to be changed. Thanks!

masatake commented:
I sent an invitation to you. You can make a repository for the snap file at universal-ctags organization after accepting the invitation. You may add a link to your new repository to README.md of universal-ctags/ctags repository. Feel free to make a pull request for updating README.md.

The name "universal-ctags" or "u-ctags" is o.k. What I strongly object is using .ctags as an file name extension. 

I will try the snap file on Fedora when I get time.
Thank you.

masatake commented:
I'm very confused when I see your new repo.
I assumed you can maintain only snapfile like universal-ctags/ctags-docker just maintains Dockerfile.
Howeveer, your new repository includes all ctags source code. Do you need all the files?

As far as reading snapcraft.yaml, the file has a field for the git repository for the target source code.
```
parts:
  universal-ctags:
    source-type: git
    source: https://github.com/universal-ctags/ctags.git
    # See 'snapcraft plugins'
    plugin: autotools
```

So I guessed the snap tool chain has ability to fetch source code from external repository.

bboymimi commented:
Thank you for the comment, I'm also now testing the procedure on my own GitHub branch and will figure out soon.
